src/System/refine.rs

```diff
@@ -17,47 +19,42 @@ pub fn check_refinement(
     sys2.initialize(dimensions);
     println!("Dimensions {}", dimensions);
     //Firstly we check the preconditions
-    if !check_preconditions(&sys1, &sys2, dimensions) {
+    if !check_preconditions(&sys1, &sys2, dimensions)? {
         println!("preconditions failed - refinement false");
-        return Ok(false);
+        return Ok(Ok(false));
     }
 
-    let inputs = sys2.get_input_actions();
-    let outputs = sys1.get_output_actions();
+    let inputs = sys2.get_input_actions()?;
+    let outputs = sys1.get_output_actions()?;
 
     let initial_locations_1 = sys1.get_initial_location();
     let initial_locations_2 = sys2.get_initial_location();
 
     if initial_locations_1 == None {
-        return Ok(initial_locations_2 == None);
+        return Ok(Ok(initial_locations_2 == None));
     }
 
     if initial_locations_2 == None {
-        return Ok(false); //The empty automata cannot implement
+        return Ok(Err(String::from(
```

Comment:
Why is this not just Ok(false) or Ok(Ok(false)), it is not an exception. If we want to carry the reason why the refinement failed, i think we should do it everywhere we return false and probably not with Result, but instead some custom type, or at least encapsulate the Result object better.